elastic/kibana
# Report CSP violations using EBT

Original POC issue: #153584
Part of issue: [#162974](https://github.com/elastic/kibana/issues/162974)

## Summary

This PR adds the ability to capture CSP violations using EBT:

- Adds `POST /internal/security/analytics/_record_violations` endpoint.
- Adds `report-sample` directive to default CSP header. 
- Adds Hapi `override` option to route config. (Explanation here https://github.com/elastic/kibana/pull/159489/files#diff-6cc0985273a2c3186d12e916aed324c437e996143e47898765ba95d7fbef0375R145-R166)

## Testing

1. Update `kibana.dev.yml` config:

```yml
logging.appenders:
  console:
    type: console
    layout:
      type: json
logging.loggers:
  - name: analytics
    level: all
    appenders: [console]

server.customResponseHeaders.Reporting-Endpoints: violations-endpoint="https://localhost:5601/xyz/internal/security/analytics/_record_violations"
server.customResponseHeaders.Content-Security-Policy-Report-Only: style-src none; report-to violations-endpoint
csp.report_to: [violations-endpoint]
```

2. Run ES with SSL: `yarn es snapshot --license trial --ssl`
4. Run Kibana with SSL: `yarn start --ssl`
5. Quit Chrome if already running
6. Open Chrome using: `/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --ignore-certificate-errors --ignore-urlfetcher-cert-requests --short-reporting-delay "https://localhost:5601/"`
7. Observe telemetry events being reported in the Kibana process:

```json
{
  "@timestamp": "2023-06-12T15:00:17.794+01:00",
  "message": "Report event \"security_csp_violation\"",
  "ebt_event": {
    "timestamp": "2023-06-12T14:00:17.794Z",
    "event_type": "security_csp_violation",
    "context": {},
    "properties": {
      "created": "1686578417796",
      "url": "https://localhost:5601/dev/login?next=%2Fdev%2F",
      "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36",
      "blockedURL": "inline",
      "columnNumber": 11,
      "disposition": "report",
      "documentURL": "https://localhost:5601/dev/login?next=%2Fdev%2F",
      "effectiveDirective": "style-src-elem",
      "lineNumber": 353,
      "originalPolicy": "style-src none; report-to violations-endpoint",
      "referrer": "",
      "sample": "",
      "sourceFile": "https://localhost:5601/dev/9007199254740991/bundles/plugin/security/1.0.0/security.chunk.0.js",
      "statusCode": 200
    }
  }
}
```
